Add significant optional debug logging for MSAA events

michaelDCurran opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
None.

### Summary of the issue:
There is not much logging for the receiving and processing of MSAA `winEvents` in NVDA. It is currently hard to debug issues around missing MSAA events and or creating IAccessible objects from events.

### Description of how this pull request fixes the issue:
Ads a significant amount of debug and debugWarning messages to the log in the `winEvent` hook, the limiter, and the various `process*WinEvent` functions. **This logging must be turned on  with the MSAA debug log checkbox in NVDA's advanced settings.**
When `winEvent` params are logged, the `eventID` and `objectID` constant names are resolved, as well as the window class name for the window handle.

### Testing performed:
Ran NVDA with and without this logging turned on.

### Known issues with pull request:
None.

### Change log entry:
Changes for developers:
- Significant debug logging for MSAA events can be now enabled in NVDA's Advanced settings.

michaelDCurran committed:
Add significant optional debug logging for MSAA events.

leonardder commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,45 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
+			log.debug("Dropping winEvent for invalid window")
 		return None
 	# Make sure this window does not have a ghost window if possible
 	if NVDAObjects.window.GhostWindowFromHungWindow and NVDAObjects.window.GhostWindowFromHungWindow(window):
```
If I"m correct, GhostWindowFromHungWindow was introduced in Windows 7, so the try except and the boolean value check can be removed

michaelDCurran replied:
I think I'd prefer to leave GhostWindowFromHungWindow for this particular pr, as that win32 function does not seem to actually be documented officially by Microsoft as far as I can see... so don't want to change any assumption about where it is available.

leonardder commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,45 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 	# We do not support MSAA object proxied from native UIA
 	if UIAHandler.handler and UIAHandler.handler.isUIAWindow(window):
+		if isMSAADebugLoggingEnabled():
+			log.debug("Dropping winEvent for native UIA window")
```
```suggestion
			log.debug(f"Dropping winEvent for native UIA window {window}")
```

leonardder commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,45 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 		# See if we already know an object by this win event info
 		obj = liveNVDAObjectTable.get((window, objectID, childID), None)
+		if isMSAADebugLoggingEnabled() and obj:
+			log.debug("Fetched existing NVDAObject for winEvent from liveNVDAObjectTable")
```
```suggestion
			log.debug(f"Fetched existing NVDAObject for winEvent from liveNVDAObjectTable: : {getWinEventLogInfo(window, objectID, childID)}")
```

leonardder commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,45 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 	# At this point if we don't have an object then we can't do any more
 	if not obj:
+		if isMSAADebugLoggingEnabled():
+			log.debug("Could not instantiate an NVDAObject for winEvent")
```
```suggestion
			log.debug(f"Could not instantiate an NVDAObject for winEvent: {getWinEventLogInfo(window, objectID, childID, eventID)}")
```

LeonarddeR reviewed (commented):
I think overall this looks good, but may be it is good to ensure that most logging items log the event info as well, as I suggested in several comments. Or will it always be clear what the source event is from the order of log entries?

codeofdusk commented:
In the changelog entry, swap the word order:

* Significant debug logging for MSAA events can now be enabled in NVDA's Advanced settings.

feerrenrut commented:
One thing that might make the conditional logging for this a little less repetitive is to install a new log handler, for an example see:
`log._speechManagerDebug` in `source/speech/manager.py:88`

One problem with this approach is that it introduces a hidden dependency for other files wishing to use this. Perhaps these special logging categories should be added / installed by logHandler.

michaelDCurran committed:
Fix typo

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

michaelDCurran commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -99,6 +99,57 @@
+		return f"window {window} ({windowClassName}), objectID {objectIDName} and childID {childID}"
+
+
+def isMSAADebugLoggingEnabled():
```
IAccessible really is just the name of one of the MSAA COM interfaces. You could possibly argue that we should have called IAccessibleHandler MSAAHandler. But I don't think it is worth the change now.

michaelDCurran committed:
Include winEvent params in log message.

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

michaelDCurran committed:
Include winEvent params in log message.

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

michaelDCurran committed:
Include window handle and windowClass name in log message about native UIA window.

Co-authored-by: Leonard de Ruijter <leonardder@users.noreply.github.com>

michaelDCurran committed:
Fix linting issues.

michaelDCurran commented:
@feerrenrut  I think I'd prefer to leave this as explicit if-checks as there is code that may be costly if run, even if the string itself was not printed to the log.

michaelDCurran commented:
@leonardder  or @feerrenrut   is there any feedback you still need me to address here?

leonardder commented on source/IAccessibleHandler/internalWinEventHandler.py:
```diff
@@ -114,14 +127,20 @@ def winEventCallback(handle, eventID, window, objectID, childID, threadID, times
 			return
+		if isMSAADebugLoggingEnabled():
+			log.debug(
+				f"Adding winEvent to limitor: {getWinEventLogInfo(window, objectID, childID, eventID)}"
```
limitOr should be limitEr

```suggestion
				f"Adding winEvent to limiter: {getWinEventLogInfo(window, objectID, childID, eventID)}"

```

LeonarddeR reviewed (dismissed):
Apart from one typo, this looks ok

feerrenrut commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,45 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 	# Ignore any events with invalid window handles
 	if not window or not winUser.isWindow(window):
+		if isMSAADebugLoggingEnabled():
+			log.debug("Dropping winEvent for invalid window")
```
I think @leonardder's suggestion here is worth adding.

feerrenrut commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -522,27 +573,51 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 	# Make sure this window does not have a ghost window if possible
 	if NVDAObjects.window.GhostWindowFromHungWindow and NVDAObjects.window.GhostWindowFromHungWindow(window):
+		if isMSAADebugLoggingEnabled():
+			log.debug("Dropping winEvent for ghosted hung window")
```
Might be worth adding which window this message is for?

feerrenrut commented on source/IAccessibleHandler/internalWinEventHandler.py:
```diff
@@ -162,18 +181,33 @@ def _shouldGetEvents():
 			# Either the foreground window is now correct
 			# or we've already had the maximum number of defers.
 			# (Sometimes, foreground events are fired even when the foreground hasn't actually changed.)
+			if curForegroundWindow != _deferUntilForegroundWindow:
+				log.debugWarning(
+					"Foreground took too long to change. "
+					f"Foreground still {curForegroundWindow} ({curForegroundClassName}). "
+					f"Should be {_deferUntilForegroundWindow} ({futureForegroundClassName})"
+				)
```
Note, this isn't protected by `isMSAADebugLoggingEnabled`, however I think that is appropriate for this message.

feerrenrut reviewed (changes requested):
It would be good if some notes were added about this to highlight the setting and messages to look for in certain use-cases. Perhaps to the developer documentation?

michaelDCurran committed:
* Included winEvent info to several more log calls.
* Added several more log calls.
* Improved log call for error in AccessibleObjectFromEvent and put it behind isMSAADebugLoggingEnabled.
* Fixed typo.
* Included threadID, processID and process name in winEvent log info where possible.

michaelDCurran commented:
I'd prefer to get this implementation merged before spending any time on possible dev docs, as this code could become out of date rather quickly, and  this logging has already proved extremely useful in the last little while.

At very least though, I will put here a quick explanation. Someone else is welcome to try and incorporate this into the dev guide before I get around to it:

## MSAA Logging ##
It may be useful to see just how NVDA receives and processes events for Microsoft Active Accessibility (MSAA), in order to understand why a particular focus event is being ignored, or why NVDA is extra slow etc.

You can turn on logging of MSAA winEvents by setting NVDA's log level to debug, and also turnning on the MSAA checkbox in the debug logging section of NVDA's Advanced settings panel.

With this logging switched on, you will now see particular messages logged such as the following:
```
Hook received winEvent: EVENT_OBJECT_FOCUS, window 1443880 (Button), objectID OBJID_CLIENT, childID 0, from process 5512 (explorer), thread 21136
```
This is the point where NVDA first receives the winEvent directly from Windows. No processing or filtering has happened at all yet. This log message shows that the app in question has actually fired the winEvent.

```
DEBUG - IAccessibleHandler.internalWinEventHandler.winEventCallback (10:54:56.442) - MainThread (32):
Adding winEvent to limiter: EVENT_OBJECT_FOCUS, window 1443880 (Button), objectID OBJID_CLIENT, childID 0, from process 5512 (explorer), thread 21136
```
Here NvDA is adding the winEvent to its winEventLimiter, to ensure that we do not get flooded with winEvents. The limiter tries to filter out duplicate winEvents, and also limits the number of events for a thread, per NVDA core cycle.
It is possible that a winEvent will be dropped before this point, due to the window being invalid, or perhaps it is a locationChange event for an object other than the caret (which we have no use for). If a winEvent is dropped before this point, another specific message will be logged.

```
DEBUG - IAccessibleHandler.orderedWinEventLimiter.OrderedWinEventLimiter.flushEvents (10:54:56.458) - MainThread (32):
Emitting winEvent EVENT_OBJECT_FOCUS, window 1443880 (Button), objectID OBJID_CLIENT, childID 0, from process 5512 (explorer), thread 21136
```
Here the winEvent is coming out of the limiter on the next NvDA core cycle, and will be processed via one or more specific winEvent handling functions, and either dropped or queued as an NVDA event.
For example, the following messages show that the focus winEvent is being handled by processFocusWinEvent, and in in tern converted from a winEvent to an NVDA event:
```
DEBUG - IAccessibleHandler.processFocusWinEvent (12:10:25.854) - MainThread (3636):
Processing focus winEvent: window 395144 (Button), objectID OBJID_CLIENT, childID 0, process 5512 (explorer), force False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (12:10:25.854) - MainThread (3636):
Creating NVDA event from winEvent: EVENT_OBJECT_FOCUS, window 395144 (Button), objectID OBJID_CLIENT, childID 0, process 5512 (explorer), use cache False
DEBUG - IAccessibleHandler.winEventToNVDAEvent (12:10:25.854) - MainThread (3636):
winEvent mapped to NVDA event: gainFocus
DEBUG - IAccessibleHandler.winEventToNVDAEvent (12:10:25.854) - MainThread (3636):
Successfully created NvDA event gainFocus for <NVDAObjects.IAccessible.Button object at 0x04C78330> from winEvent EVENT_OBJECT_FOCUS, window 395144 (Button), objectID OBJID_CLIENT, childID 0, process 5512 (explorer)
```

Sometimes a winEvent is for an NvDAObject that is already being tracked by NVDA, such as the current focus. In this case, the winEvent will be redirected straight to the existing NVDAObject, rather than creating a new one with AccessibleObjectFromEvent etc:
```
DEBUG - IAccessibleHandler.winEventToNVDAEvent (12:13:33.703) - MainThread (19248):
Fetched existing NVDAObject <NVDAObjects.IAccessible.Button object at 0x08053310> from liveNVDAObjectTable for winEvent window 5636926 (Button), objectID OBJID_CLIENT, childID 0, process 5512 (explorer)
```

There are many reasons why a winEvent may be deliberately dropped along the way. 
  One might be that the limiter detected too many events for a particular thread during an NVDA core cycle. If so you will see this message:
```
DEBUG - IAccessibleHandler.orderedWinEventLimiter.OrderedWinEventLimiter.flushEvents (12:16:24.886) - MainThread (19248):
winEvent limit for thread 17008 hit for this core cycle
```

Or perhaps, the winEvent was for a window that supported UI automation natively:
```
DEBUG - IAccessibleHandler.winEventToNVDAEvent (12:17:44.604) - MainThread (19248):
Native UIA window. Dropping winEvent EVENT_OBJECT_STATECHANGE, window 854612 (DirectUIHWND), objectID 100794482, childID 0, process 5512 (explorer)
```



DEBUG - IAccessibleHandler.processGenericWinEvent (10:54:56.458) - MainThread (32):
Directing winEvent to focus object

josephsl commented:
Hi,

I guess it can go under "events" section where we describe list of common events. Due to the nature of this PR, I think a section on debugging API level events might be useful, which would then be linked from the events section.

Below is what Windows 10 App Essentials add-on internals document says about UIA event tracking, which I think would be a good starting point:

#### Tracking UIA events for controls

The Windows 10 Objects global plugin also has ability to track UIA events for controls and log info  about them, executed via `uiaDebugLogging` function that takes an object and the event name. This function records the following if NVDA is started with debug logging enabled or told to monitor specific events and/or events from specific apps:

* What the object actually is.
* Object name.
* Name of the event being logged.
* App where the control can be found (specifically, the app module).
* Automation Id if possible.
* UIA class name.
* For controller for event, the list of objects the given control depends on.
* For tooltip open event, the GUI framework that powers the element.
* For item status event, new item status text.
* For state change event, current element status.

Thanks.

michaelDCurran committed:
Added yet more MSAA debug log messages.

michaelDCurran commented:
@feerrenrut  I think I have now addressed everything. But I have also added yet more log calls. Plus, as we quickly spoke about last week, I have now included the process name in the winEvent log info.

feerrenrut commented on source/IAccessibleHandler/__init__.py:
```diff
@@ -626,6 +644,11 @@ def winEventToNVDAEvent(eventID, window, objectID, childID, useCache=True):
 			obj = SDMChild
+	if isMSAADebugLoggingEnabled():
+		log.debug(
+			f"Successfully created NvDA event {NVDAEventName} for {obj} "
```
Minor: lower case v in NVDA

michaelDCurran committed:
Fix typo

michaelDCurran committed:
Update what's new
